File: internal/app/repo/postgres/user_repo.go
package postgres

import (
	"context"
	"database/sql"
	"fmt"
	"log/slog"
	"mahir-trade-be/internal/app/models"
	"mahir-trade-be/internal/app/repo/postgres/queries"
	"strings"

	"go.uber.org/dig"
)

type (
	UserRepo interface {
		CreateUser(ctx context.Context, req models.User) (id int, err error)
		FindUserByEmailOrUsername(ctx context.Context, req string) (user models.User, err error)
		FindUserByEmailAndUsername(ctx context.Context, email, username string) (user models.User, err error)
		GetUserByID(ctx context.Context, id int64) (user models.User, err error)
		GetUserByUUID(ctx context.Context, uuid string) (user models.User, err error)
		UpdateTypeUser(ctx context.Context, isActive bool, operator string, id int64) (typeUser bool, err error)
		GetAllUser(ctx context.Context, req models.PaginationRequest) (users []models.GetUsersBOResponse, totalCount int64, err error)
		UpdatePassword(ctx context.Context, password string, operator string, id int64) (isUpdated bool, err error)
		SetUserVerified(ctx context.Context, userId int64) (err error)
	}

	UserRepoImpl struct {
		dig.In

		*sql.DB
	}
)

func NewUserRepo(impl UserRepoImpl) UserRepo {
	return &impl
}

func (u *UserRepoImpl) CreateUser(ctx context.Context, req models.User) (id int, err error) {
	rows, err := u.QueryContext(ctx, queries.QueryCreateUser, req.Email, req.PhoneNumber, req.Username, req.Password)
	if err != nil {
		if strings.Contains(err.Error(), "duplicate key value violates unique constraint") {
			return id, fmt.Errorf("email or username already exist")
		}

		slog.ErrorContext(ctx, fmt.Sprintf("error while CreateUser err: %v", err.Error()))
		return id, err
	}

	defer rows.Close()

	for rows.Next() {
		err = rows.Scan(&id)
		if err != nil {
			return id, err
		}
	}

	return id, nil
}

func (u *UserRepoImpl) FindUserByEmailOrUsername(ctx context.Context, req string) (user models.User, err error) {
	row := u.QueryRowContext(ctx, queries.QueryFindUserByEmailOrUsename, req)
	err = row.Scan(&user.UserID, &user.UUID, &user.PhoneNumber, &user.Username, &user.Email, &user.Password, &user.VerifiedAt)
	if err != nil {
		slog.ErrorContext(ctx, fmt.Sprintf("error while FindUserByEmailOrUsername err: %v", err.Error()))
		return user, err
	}

	return user, nil
}

func (u *UserRepoImpl) GetUserByID(ctx context.Context, id int64) (user models.User, err error) {
	row := u.QueryRowContext(ctx, queries.QueryGetUserByID, id)
	err = row.Scan(&user.UserID, &user.UUID, &user.PhoneNumber, &user.Username, &user.Email, &user.Password, &user.IsActive)
	if err != nil {
		slog.ErrorContext(ctx, fmt.Sprintf("error while GetUserByID err: %v", err.Error()))

		if err == sql.ErrNoRows {
			return user, fmt.Errorf("user not found")
		}

		return user, err
	}

	return user, nil
}

func (u *UserRepoImpl) GetUserByUUID(ctx context.Context, uuid string) (user models.User, err error) {
	row := u.QueryRowContext(ctx, queries.QueryGetUserByUUID, uuid)
	err = row.Scan(&user.UserID, &user.UUID, &user.PhoneNumber, &user.Username, &user.Email, &user.Password)
	if err != nil {
		slog.ErrorContext(ctx, fmt.Sprintf("error while GetUserByUUID err: %v", err.Error()))
		return user, err
	}

	return user, nil
}

func (u *UserRepoImpl) UpdateTypeUser(ctx context.Context, isActive bool, operator string, id int64) (typeUser bool, err error) {
	_, err = u.QueryContext(ctx, queries.QueryUpdateTypeUser, isActive, operator, id)
	if err != nil {
		slog.ErrorContext(ctx, fmt.Sprintf("error while UpdateTypeUser err: %v", err.Error()))
		return typeUser, err
	}

	return true, nil
}
func (u *UserRepoImpl) FindUserByEmailAndUsername(ctx context.Context, email, username string) (user models.User, err error) {
	row := u.QueryRowContext(ctx, queries.QueryFindUserByEmailAndUsername, email, username)
	err = row.Scan(&user.UserID, &user.UUID, &user.PhoneNumber, &user.Username, &user.Email, &user.Password)
	if err != nil {
		slog.ErrorContext(ctx, fmt.Sprintf("error while FindUserByEmailAndUsername err: %v", err.Error()))
		return user, err
	}

	return user, nil
}

func (u *UserRepoImpl) GetAllUser(ctx context.Context, req models.PaginationRequest) (users []models.GetUsersBOResponse, totalCount int64, err error) {
	query := queries.QueryGetUsers
	queryParams := []interface{}{}

	if req.Search != "" {
		// currently only can find by email
		query += " AND u.email ILIKE '%' || $1 || '%'"
		queryParams = append(queryParams, strings.ToLower(req.Search))
	}

	if req.MembershipStatus != "" {
		filterQuery := fmt.Sprintf(" AND um.status = $%d", len(queryParams)+1)
		switch strings.ToUpper(req.MembershipStatus) {
		case string(models.MembershipStatusActive):
			filterQuery = fmt.Sprintf(" AND um.is_membership_active = $%d", len(queryParams)+1)
			queryParams = append(queryParams, true)
		case string(models.MembershipStatusPreOrder):
			queryParams = append(queryParams, models.MembershipStatusPreOrder)
		case string(models.MembershipStatusExpired):
			queryParams = append(queryParams, models.MembershipStatusExpired)
		}
		query += filterQuery
	}

	if req.SortBy == "" {
		req.SortBy = "DESC"
	}

	sortDirection := strings.ToUpper(req.SortBy)
	if sortDirection != "ASC" && sortDirection != "DESC" {
		sortDirection = "DESC"
	}

	query += fmt.Sprintf(" ORDER BY u.created_at %s LIMIT $%d OFFSET $%d", sortDirection, len(queryParams)+1, len(queryParams)+2)
	queryParams = append(queryParams, req.Limit, int((req.Page-1)*req.Limit))

	rows, err := u.QueryContext(ctx, query, queryParams...)
	if err != nil {
		slog.ErrorContext(ctx, fmt.Sprintf("[repo][GetAllUsers] while QueryContext, err: %v", err.Error()))
		return
	}

	defer rows.Close()

	for rows.Next() {
		var user models.GetUsersBOResponse
		err = rows.Scan(
			&totalCount,
			&user.UserID,
			&user.UUID,
			&user.PhoneNumber,
			&user.Email,
			&user.Username,
			&user.IsActive,
			&user.AccountType,
			&user.MembershipExpiredData,
			&user.CreatedAt,
			&user.CreatedBy,
		)
		if err != nil {
			slog.ErrorContext(ctx, fmt.Sprintf("[repo][GetAlltUser] while Scan, err: %v", err.Error()))
			return
		}

		users = append(users, user)
	}

	return
}

func (u *UserRepoImpl) UpdatePassword(ctx context.Context, password string, operator string, id int64) (isUpdated bool, err error) {
	row, err := u.ExecContext(ctx, queries.QueryUpdatePassword, password, operator, id)
	if err != nil {
		slog.ErrorContext(ctx, fmt.Sprintf("[repo][UpdatePassword] while ExecContext, err: %v", err.Error()))
		return
	}

	affected, err := row.RowsAffected()
	if err != nil {
		slog.ErrorContext(ctx, fmt.Sprintf("[repo][UpdatePassword] while RowsAffected, err: %v", err.Error()))
		return
	}

	isUpdated = affected > 0
	return

}

func (u *UserRepoImpl) SetUserVerified(ctx context.Context, userId int64) (err error) {
	_, err = u.ExecContext(ctx, queries.QuerySetUserVerified, userId)
	if err != nil {
		slog.ErrorContext(ctx, fmt.Sprintf("[repo][SetUserVerified] while ExecContext, err: %v", err.Error()))
		return
	}

	return
}

File: src/repo/user_repo.ts
import { UserQueries } from "./queries/user_queries";
import { User, GetUsersBOResponse } from "../models/user.model";
import db from "../db/db.config";
import { PaginationRequest } from "../utils/response";

export class UserRepository {
  async createUser(user: User): Promise<number> {
    try {
      const result = await db.query(UserQueries.CreateUser, [
        user.email,
        user.phone_number,
        user.username,
        user.password,
      ]);
      return result.rows[0].user_id;
    } catch (err: any) {
      if (err.message.includes("duplicate key value")) {
        throw new Error("email or username already exist");
      }
      throw err;
    }
  }

  async findUserByEmailOrUsername(identity: string): Promise<User | null> {
    console.log("Identity data: ", identity);

    const result = await db.query(UserQueries.FindUserByEmailOrUsername, [
      identity,
    ]);

    console.log("Result Hasil: ", result);

    return result.rows[0] || null;
  }

  async findUserByEmailAndUsername(
    email: string,
    username: string
  ): Promise<User | null> {
    const result = await db.query(UserQueries.FindUserByEmailAndUsername, [
      email,
      username,
    ]);
    return result.rows[0] || null;
  }

  async getUserById(id: number): Promise<User> {
    const result = await db.query(UserQueries.GetUserByID, [id]);
    if (result.rowCount === 0) {
      throw new Error("user not found");
    }
    return result.rows[0];
  }

  async getUserByUUID(uuid: string): Promise<User | null> {
    const result = await db.query(UserQueries.GetUserByUUID, [uuid]);
    return result.rows[0] || null;
  }

  async updateTypeUser(
    id: number,
    isActive: boolean,
    operator: string
  ): Promise<boolean> {
    const result = await db.query(UserQueries.UpdateTypeUser, [
      isActive,
      operator,
      id,
    ]);

    // rowCount pasti ada, defaultnya 0 kalau tidak ada row yang kena update
    return result.rowCount! > 0;
  }

  async getAllUser(
    req: PaginationRequest
  ): Promise<{ users: any[]; totalCount: number }> {
    let query = UserQueries.GetUsers;
    const params: any[] = [];

    // 🔍 Filter search by email
    if (req.search) {
      query += ` AND LOWER(u.email) ILIKE '%' || $${params.length + 1} || '%'`;
      params.push(req.search.toLowerCase());
    }

    // 🧩 Filter membership status
    if (req.membership_status) {
      const status = req.membership_status.toUpperCase();

      if (status === "ACTIVE") {
        query += ` AND um.is_membership_active = $${params.length + 1}`;
        params.push(true);
      } else if (status === "EXPIRED") {
        query += ` AND um.is_membership_active = $${params.length + 1}`;
        params.push(false);
      } else if (status === "PRE_ORDER") {
        query += ` AND um.is_membership_active IS NULL`;
      }
    }

    // 🔢 Sort + Pagination (⚠️ tambahin spasi di depan ORDER)
    const sortDirection = req.sort_by?.toUpperCase() === "ASC" ? "ASC" : "DESC";
    query += ` ORDER BY u.created_at ${sortDirection} LIMIT $${
      params.length + 1
    } OFFSET $${params.length + 2}`;
    params.push(req.limit, (req.page - 1) * req.limit);

    // Debug log
    console.log("[DEBUG] Final Query:", query);
    console.log("[DEBUG] Params:", params);

    // 🧠 Execute query
    const result = await db.query(query, params);

    // 🧾 Map data result
    const users = result.rows.map((row: any) => ({ ...row }));
    const totalCount =
      result.rows.length > 0 ? parseInt(result.rows[0].total_count, 10) : 0;

    return { users, totalCount };
  }

  async updatePassword(
    userId: number,
    password: string,
    operator: string
  ): Promise<boolean> {
    const result = await db.query(UserQueries.UpdatePassword, [
      password,
      operator,
      userId,
    ]);
    return result.rowCount! > 0;
  }

  async setUserVerified(userId: number): Promise<boolean> {
    const result = await db.query(UserQueries.SetUserVerified, [userId]);
    return result.rowCount! > 0;
  }
}
